text/0074-global-init.md

```diff
@@ -0,0 +1,176 @@
+following strategies.
+
+### Service provider mechanism
+
+Where the language provides a commonly accepted way to inject SDK
+components, it should be preferred.  The Java SPI (Service Provider
+Interface) supports loading and configuring the SDK before it is first
+used.  This kind of support is the preferred choice in languages with
+common support for automatic dependency injection.
+
+In this case, there is no use-before-configuration question to
+address, as there is in some languages with support for static
+initialization.
+
+### Explicit initializer mechanism
+
+When it is not possible to ensure the SDK is installed and configured
+before the API is first used, initializing the default, global SDK
+instances becomes the user's responsibility.
+
+Methods to set the global instances shall be independent, allowing
+each SDK component to be intialized separately when the process
+starts.  The methods shall be declared in a separate API package,
+e.g., named `global`:
+
+```golang
+package global
+
+import (
+    "go.opentelemetry.io/otel/api/context/propagation"
+    "go.opentelemetry.io/otel/api/metric"
+    "go.opentelemetry.io/otel/api/trace"
+)
+
+// SetMeterProvider initializes the global Meter provider.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Meter provider, the default global
+// Meter provider acts as a "forwarding" SDK.  The default global
+// Meter provider will begin forwarding to the installed Meter
+// provider once this is called.
+func SetMeterProvider(metric.Provider) { ... }
+
+// SetPropagators initializes the global Propagators instance.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Meter provider, the default global
+// Propagators instance performs pass-through W3C Traceparent and
+// Correlation-Context propagation.
+func SetPropagators(propagation.Propagators) { ... }
+
+// SetPropagators initializes the global Tracer provider.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Tracer provider, the default global
+// Tracer provider acts as a "forwarding" SDK.  The default global
+// Tracer provider will begin forwarding to the installed Tracer
+// provider once this is called.
+func SetTraceProvider(trace.Provider) { ... }
+```
+
+#### Requirements
+
+We anticipate third party libraries using the global instances before
+they are installed, and we wish for references obtained through these
+instances to become functional once the corresponding implementation
+is initialized.  The default instances returned by the global getters
+for Tracer provider, Meter provider, and Propagators must "forward" to
+the real SDK implementation once it is installed.
+
+#### Tracer
+
+Tracers obtained through the provider will become functional when the
+user's Tracer SDK is installed as the global instance.
```

Comment:
This is a very interesting requirement. So, if I read this correctly, every Tracer/Meter must be a shim with the capability of having it's underlying implementation swapped out at runtime?

Reply:
That's the idea, yes. This OTEP was originally written before the named Tracer/Meter was introduced, so the text is a bit light on this point, but I believe that if the global instance can be used before initialization, then this type of shim is needed to properly support it. 

It's certainly a pattern in metrics APIs to allocate instruments statically, so I was mostly focused on that support. I later realized that named Tracer and named Meter would require the same thing, otherwise you'd have applications starting up with a bunch of No-op implementations in static variables.

Reply:
Yes. The more I think about the crazy Java ecosystem, the more I think we're going to need something that can be swapped out at runtime, post class-loading. Trying to anticipate all classloader initialization ordering and making sure that the SDK can pre-emptively initialize the globals feels like a fools-errand. Building the dynamism into the SDk from the beginning will be much simpler, even if it's a little more complex than it needs to be on the surface.

Reply:
Btw, the (old) OT Java global `Tracer` is a singleton that can be swapped once, and simply forwards its calls (and yes, it worked just fine all this time ;) ) 

https://github.com/opentracing/opentracing-java/blob/master/opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java

Reply:
Oh! I did not know that. It's nice to see a precedent. I'll add this detail.